File: BOJ/Silver/BOJ9012.py
# https://www.acmicpc.net/problem/9012

import sys
input = sys.stdin.readline
n = int(input())

for i in range(n) :
    ps = list(input().rstrip())
    answer = []
    flag = 0

    for lps in ps :
        if lps == "(" :
            answer.append(lps)
        else :
            if len(answer) != 0:
                answer.pop()
            else :
                answer.append(lps)
                break

    if len(answer) == 0:
        print("YES")
    else :
        print("NO")

File: BOJ/Silver/BOJ9012.java
/* https://www.acmicpc.net/problem/9012 */

import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        Stack<String> stack = null;
        String prev = null;
        String answer = null;

        for(int i=0; i<n; i++){
            String[] target = br.readLine().split("");
            stack = new Stack<>();

            for(int j=0; j<target.length; j++){
                if(target[j].equals("(")){
                    stack.push(target[j]);
                } else {
                    if(stack.size() >= 1){
                        prev = stack.peek();
                        if(prev.equals("(")){
                            stack.pop();
                        }
                    } else {
                        stack.push(target[j]);
                    }
                }
            }
            answer = stack.size() == 0 ? "YES" : "NO";
            System.out.println(answer);
        }
    }
}